In [2]:
import pandas as pd
import numpy as np
import matplotlib as mp
import seaborn as sn
import csv

In [3]:
brands = pd.read_csv('brands_v2.csv')
finance = pd.read_csv('finance.csv')
info = pd.read_csv('info_v2.csv')
reviews = pd.read_csv('reviews_v2.csv')
traffic = pd.read_csv('traffic_v3.csv')

## Data Layout:

- **Brands**
    - product_id*
    - brand
- **Finance**
    - product_id*
    - listing_price
    - sale_price
    - discount
    - revenue
- **Info**
    - product_name
    - product_id**
    - description
- **Reviews**
    - product_id*
    - rating
    - review
- **Traffic**
    - product_id*
    - last_visited

In [94]:
print("Brands Dataset Info")
brands.info()

x = brands.head()
print('\n')
print("Sample Data")
print(x)

Brands Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3179 entries, 0 to 3178
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   product_id  3179 non-null   object
 1   brand       3120 non-null   object
dtypes: object(2)
memory usage: 49.8+ KB


Sample Data
  product_id   brand
0     AH2430     NaN
1     G27341  Adidas
2     CM0081  Adidas
3     B44832  Adidas
4     D98205  Adidas


In [95]:
print("Finance Dataset Info")
finance.info()

x = finance.head()
print('\n')
print("Sample Data")
print(x)

Finance Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3179 entries, 0 to 3178
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     3179 non-null   object 
 1   listing_price  3120 non-null   float64
 2   sale_price     3120 non-null   float64
 3   discount       3120 non-null   float64
 4   revenue        3120 non-null   float64
dtypes: float64(4), object(1)
memory usage: 124.3+ KB


Sample Data
  product_id  listing_price  sale_price  discount  revenue
0     AH2430            NaN         NaN       NaN      NaN
1     G27341          75.99       37.99       0.5  1641.17
2     CM0081           9.99        5.99       0.4   398.93
3     B44832          69.99       34.99       0.5  2204.37
4     D98205          79.99       39.99       0.5  5182.70


In [96]:
print("Info Dataset Info")
info.info()

x = info.head()
print('\n')
print("Sample Data")
print(x)

Info Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3179 entries, 0 to 3178
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_name  3120 non-null   object
 1   product_id    3179 non-null   object
 2   description   3117 non-null   object
dtypes: object(3)
memory usage: 74.6+ KB


Sample Data
                                       product_name product_id  \
0                                               NaN     AH2430   
1              Women's adidas Originals Sleek Shoes     G27341   
2                 Women's adidas Swim Puka Slippers     CM0081   
3  Women's adidas Sport Inspired Questar Ride Shoes     B44832   
4          Women's adidas Originals Taekwondo Shoes     D98205   

                                         description  
0                                                NaN  
1  A modern take on adidas sport heritage, tailor...  
2  These adidas Puka slippers for women's come wi

In [97]:
print("Reviews Dataset Info")
reviews.info()

x = reviews.head()
print('\n')
print("Sample Data")
print(x)

Reviews Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3179 entries, 0 to 3178
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   product_id  3179 non-null   object 
 1   rating      3120 non-null   float64
 2   reviews     3120 non-null   float64
dtypes: float64(2), object(1)
memory usage: 74.6+ KB


Sample Data
  product_id  rating  reviews
0     AH2430     NaN      NaN
1     G27341     3.3     24.0
2     CM0081     2.6     37.0
3     B44832     4.1     35.0
4     D98205     3.5     72.0


In [98]:
print("Traffic Dataset Info")
traffic.info()

x = traffic.head()
print('\n')
print("Sample Data")
print(x)

Traffic Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3179 entries, 0 to 3178
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    3179 non-null   object
 1   last_visited  2928 non-null   object
dtypes: object(2)
memory usage: 49.8+ KB


Sample Data
  product_id         last_visited
0     AH2430  2018-05-19 15:13:00
1     G27341  2018-11-29 16:16:00
2     CM0081  2018-02-01 10:27:00
3     B44832  2018-09-07 20:06:00
4     D98205  2019-07-18 15:26:00


There are 3179 total entries, with product_id not having any nulls. *description* in the *reviews* table has 3117 nulls and last_visited in traffic table has 2928 nulls. The rest of the columns in all tables have 3120 nulls.

*product_id* is the primary key for each table besides the *info* table where the *product_name* is the primary key and the *product_id* is the foreign key.

***Brands***:
- Top brands?
- Top products?

***Finance***:
- Most expensive vs. least expensive products?
- Highest sales/discount?
- Highest revenue?

***Info***:
- Most vs. least amount of products?

***Reviews***:
- Best vs. worst rating?
- Most vs. least amount of ratings?

***Traffic***:
- Which hours and/or days is the store the busiest?
- Which products are most vs. least popular at these times?

***Joins***:
1. Join the finance and traffic table to find out **when** the store makes the most money and/or is most busy.
2. Join the finance, brands, and info table to find out **what** products bring in the most money and/or are bought the most.
3. Join the info and reviews table to find out **which** products get the **best** reviews.
    - **Worst** reviews and why?

In [30]:
from sqlalchemy import create_engine
import psycopg2

from urllib.parse import quote_plus
password = quote_plus("Rowan1016!")  # Encodes special characters
DATABASE_URL = f"postgresql://postgres:{password}@localhost:5432/sports_retail"

try:
    engine = create_engine(DATABASE_URL)
    conn = engine.connect()
    print("Connection successful!")
    conn.close()
except Exception as e:
    print("Connection failed:", e)

Connection successful!


In [31]:
%reload_ext sql
%sql $DATABASE_URL

In [32]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'  # Change to html style or try others

In [7]:
brands.to_sql("brands", engine, if_exists='replace', index=False)
finance.to_sql("finance", engine, if_exists='replace', index=False)
info.to_sql("info", engine, if_exists='replace', index=False)
reviews.to_sql("reviews", engine, if_exists='replace', index=False)
traffic.to_sql("traffic", engine, if_exists='replace', index=False)

179

In [103]:
%%sql

SELECT * FROM traffic
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sports_retail
10 rows affected.


[('AH2430', '2018-05-19 15:13:00'),
 ('G27341', '2018-11-29 16:16:00'),
 ('CM0081', '2018-02-01 10:27:00'),
 ('B44832', '2018-09-07 20:06:00'),
 ('D98205', '2019-07-18 15:26:00'),
 ('B75586', '2019-01-30 12:09:00'),
 ('CG4051', '2019-03-22 16:36:00'),
 ('CM0080', '2019-03-10 01:46:00'),
 ('B75990', '2018-05-29 08:16:00'),
 ('EE5761', '2019-11-29 17:22:00')]

In [104]:
%%sql

DELETE FROM brands
WHERE (product_id IS NULL OR brand IS NULL);

DELETE FROM finance
WHERE (product_id IS NULL OR listing_price IS NULL OR sale_price IS NULL OR discount IS NULL OR revenue IS NULL);

DELETE FROM info
WHERE (product_id IS NULL OR product_name IS NULL OR description IS NULL);

DELETE FROM reviews
WHERE (product_id IS NULL OR rating IS NULL OR reviews IS NULL);

DELETE FROM traffic
WHERE (product_id IS NULL OR last_visited IS NULL);


 * postgresql://postgres:***@localhost:5432/sports_retail
59 rows affected.
59 rows affected.
62 rows affected.
59 rows affected.
251 rows affected.


[]

# Brands:

Top brands?

In [105]:
%%sql

WITH CTE as (
    SELECT brand, COUNT(brand) AS Brand_Rank
    FROM brands
    GROUP BY brand
)
SELECT brand, Brand_Rank
FROM CTE
WHERE Brand_Rank = (SELECT MAX(Brand_Rank) FROM CTE)
UNION ALL
SELECT brand, Brand_Rank
FROM CTE
WHERE Brand_Rank = (SELECT MIN(Brand_Rank) FROM CTE);

 * postgresql://postgres:***@localhost:5432/sports_retail
2 rows affected.


[('Adidas', 2575), ('Nike', 545)]

Number of unique products?

In [29]:
%%sql

SELECT COUNT(DISTINCT(product_id))
FROM brands; 

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(3179,)]

# Finance

- Most expensive vs. least expensive products?
- Highest sales/discount?
- Highest revenue?


In [107]:
%%sql

WITH MAX_PRICE AS (
    SELECT product_id, listing_price   
    FROM financE
    ORDER BY listing_price DESC
    LIMIT 1
),
MIN_PRICE AS (
    SELECT product_id, listing_price       
    FROM finance
    WHERE NOT listing_price = '0.0'
    ORDER BY listing_price ASC
    LIMIT 1
)
SELECT MAX_PRICE.product_id AS max_price_product, MAX_PRICE.listing_price AS Max_price, MIN_PRICE.product_id AS min_price_product, 
       MIN_PRICE.listing_price AS min_price
FROM MAX_PRICE
CROSS JOIN MIN_PRICE;


 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[('BD7927', 299.99, 'CL0233', 8.99)]

In [33]:
%%sql

WITH CTE AS (
    SELECT product_id, listing_price       
    FROM finance
    WHERE listing_price = '0.0'
)
SELECT COUNT(*) as No_Price_CNT
FROM CTE;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(354,)]

In [109]:
%%sql

SELECT DISTINCT product_id as max_sale_product, sale_price as max_sale_price
FROM finance
ORDER BY sale_price DESC
LIMIT 1;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[('AT5673-700', 365.0)]

In [110]:
%%sql

SELECT product_id as max_rev_product, revenue as max_rev_price
FROM finance
ORDER BY revenue DESC
LIMIT 1;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[('310805-137', 64203.93)]

# Reviews

- Best vs. worst rating?
- Most vs. least amount of ratings?

In [132]:
%%sql

SELECT * FROM reviews
LIMIT 3;

 * postgresql://postgres:***@localhost:5432/sports_retail
3 rows affected.


[('G27341', 3.3, 24.0), ('CM0081', 2.6, 37.0), ('B44832', 4.1, 35.0)]

In [130]:
%%sql

WITH CTE1 AS (
    SELECT * FROM reviews
    WHERE rating = '0.0' AND reviews != '0.0'
),
CTE2 AS (
    SELECT * FROM reviews
    WHERE rating = '0.0' AND reviews = '0.0'
)
SELECT (SELECT COUNT(*) FROM CTE1) AS zero_rating_with_reviews, 
    (SELECT COUNT(*) FROM CTE2) AS zero_rating_without_reviews;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(70, 188)]

In [133]:
%%sql

WITH MAX_Rating AS (
    SELECT product_id, rating   
    FROM reviews
    ORDER BY rating DESC
    LIMIT 1
),
MIN_Review_Rating AS (
    SELECT product_id, rating       
    FROM reviews
    WHERE reviews != '0.0'
    ORDER BY rating ASC
    LIMIT 1
),
MIN_No_Review_Rating AS (
    SELECT product_id, rating       
    FROM reviews
    WHERE reviews = '0.0'
    ORDER BY rating ASC
    LIMIT 1
)
SELECT MAX_Rating.product_id AS max_rated_product, MAX_Rating.rating AS Max_rating, MIN_Review_Rating.product_id AS min_rated_product, 
       MIN_Review_Rating.rating AS min_rating, MIN_No_Review_Rating.product_id AS min_NR_rated_product, 
       MIN_No_Review_Rating.rating AS min_NR_rating
FROM MAX_Rating
CROSS JOIN MIN_Review_Rating
CROSS JOIN MIN_No_Review_Rating;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[('EE8042', 5.0, 'CK9664', 0.0, 'EG7155', 0.0)]

In [22]:
%%sql
SELECT COUNT(*)   
FROM reviews
WHERE rating = '5.0'

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(186,)]

In [12]:
%%sql

SELECT COUNT(*) AS review_CNT
FROM reviews
WHERE reviews = '0.0' AND rating != '0.0';

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(24,)]

In [21]:
%%sql

WITH MAX_NUM AS (
    SELECT product_id, reviews as Product_Review_CNT 
    FROM reviews
    WHERE reviews IS NOT NULL
    ORDER BY Product_Review_CNT DESC
    LIMIT 1
),
MIN_NUM AS (
    SELECT product_id, reviews as Product_Review_CNT 
    FROM reviews
    ORDER BY Product_Review_CNT ASC
    LIMIT 1
)
SELECT MAX_NUM.product_id as MAX_Product, MAX_NUM.Product_Review_CNT as MAX_Product_Review_CNT,
    MIN_NUM.product_id as MIN_Product, MIN_NUM.Product_Review_CNT as MIN_Product_Review_CNT
FROM MAX_NUM
CROSS JOIN MIN_NUM;

 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[('310805-137', 223.0, 'CL7507', 0.0)]

In [10]:
%%sql

WITH CTE1 AS (
    SELECT COUNT(*) AS review_CNT
    FROM reviews
    WHERE reviews != '0.0'
),
CTE2 AS (
    SELECT COUNT(*) AS NON_review_CNT
    FROM reviews
    WHERE reviews = '0.0'
)
SELECT CTE1.review_CNT, CTE2.NON_review_CNT FROM CTE1
CROSS JOIN CTE2;


 * postgresql://postgres:***@localhost:5432/sports_retail
1 rows affected.


[(2908, 212)]

In [27]:
%%sql   
SELECT product_id, reviews as Product_Review_CNT, rating
FROM reviews
WHERE reviews IS NOT NULL
ORDER BY Product_Review_CNT DESC, rating DESC
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/sports_retail
5 rows affected.


[('310805-137', 223.0, 4.7),
 ('880848-005', 105.0, 4.3),
 ('F35560', 99.0, 5.0),
 ('CM0070', 99.0, 5.0),
 ('B37846', 99.0, 4.9)]

**Overall Rating and Review Info**:

A total of 258 products that have a *0.0* rating.

A total of 212 products with no reviews.

Out of these totalS:

- *70* products have a *0.0* rating and also have a review. (trouble products?)
- *188* products have a *0.0* rating and no reviews. (are these products newer?)
- *24* products have a rating above zero but no reviews. (are they worth a review, like small/simple products? Are the ratings high enough to not need a written review?)

**Product Info**:

- There are 186 products with a rating of 5.0 out of 5.0.
- The product with the most reviews is *310805* with 223 reviews, more than double that of the second-most reviewed product *880848-005* with 105 reviews. The top reviewed product has a rating of 4.7 and is also the product that brings in the most revenue.

# Traffic

- Which hours and/or days is the store the busiest?
- Which products are most vs. least popular at these times?

# Joins

1. Join the finance and traffic table to find out **when** the store makes the most money and/or is most busy.
2. Join the finance, brands, and info table to find out **what** products bring in the most money and/or are bought the most.
3. Join the info and reviews table to find out **which** products get the **best** reviews.
    - **Worst** reviews and why?